rustls/src/quic.rs

```diff
@@ -779,6 +780,44 @@ impl<'a> KeyBuilder<'a> {
     }
 }
 
+/// Crypto provider context required for producing QUIC initial keys
+#[derive(Clone, Copy)]
+pub struct InitialSuite {
+    pub(crate) suite: &'static Tls13CipherSuite,
+    pub(crate) quic: &'static dyn Algorithm,
+}
+
+impl InitialSuite {
+    /// Find a suitable cipher suite from the given [`CryptoProvider`].
+    pub fn from_provider(provider: &CryptoProvider) -> Option<Self> {
```

Comment:
It occurred to me after writing this (which solves the main use case from Quinn) that this trivially generalizes to a more generic `QuicSuite` if we pass a `CipherSuite` parameter in addition to the `CryptoProvider`. In fact, given that we can now break compatibility, I think there are some benefits to be had if we make this a method on `CryptoProvider` that can yield this type given a `CipherSuite`.

I think my work on updating Quinn made it clear that the current `CryptoProvider` `Quic` abstraction is kind of painful because the `quic::Algorithm` doesn't get access to the full `Tls13CipherSuite`. The rustls 0.22 abstractions are kind of painful because a `Tls13CipherSuite` carries an `Option<quic::Algorithm>` which makes the only way of representing a cipher suite that **definitely** allows QUIC a pair of both where the `Option` has been peeled off or a type like this `InitialSuite` wrapping that.

As such, I think the better abstraction here (given that we can now break compatibility again) would be to have a trait implementation that crypto providers must implement that takes a `&'static Tls13CipherSuite` and yields `Option<quic::Suite>` (where `Suite` is a renamed version of this `InitialSuite`). This would be exposed as yet another `&'static dyn QuicSuiteProvider` field in `CryptoProvider`, I think.

Reply:
I wonder if Quic is another ordinal in `SupportedCipherSuite`? Right now there is a lot of unused baggage that comes along if someone uses say `TLS13_AES_256_GCM_SHA384` with quic, such as `Tls13CipherSuite::aead_alg` (which AIUI gets instantiated but never used). I think that would be a fair significant refactoring, and implies that (eg) `QUIC_AES_256_GCM_SHA384` would need to be introduced.

Reply:
That might be practical but seems like it gets weird in terms of managing the server handshake for example, where the choice is really between 1.2 and 1.3 and not between 1.2/1.3/QUIC. I do feel like constructing a separate `QuicSuite` with the `Tls13CipherSuite` hits a decent sweet spot.

Reply:
>  I do feel like constructing a separate QuicSuite with the Tls13CipherSuite as input hits a decent sweet spot.

I'll give this a shot and we can see how it looks :+1:

Reply:
> The rustls 0.22 abstractions are kind of painful because a Tls13CipherSuite carries an Option<quic::Algorithm> which makes the only way of representing a cipher suite that definitely allows QUIC a pair of both where the Option has been peeled off or a type like this InitialSuite wrapping that.

Or, could we address this awkwardness at the `Tls13CipherSuite` level?

```rust
impl Tls13CipherSuite {
    pub fn quic(&'static self) -> Option<crate::quic::Suite> {
        self.quic
            .map(|quic| crate::quic::Suite { quic, suite: self })
    }
}
```

Reply:
> could we address this awkwardness at the Tls13CipherSuite level?

Here's some progress on this front for consideration:

* [`cpu-quic-support-review`](https://github.com/rustls/rustls/compare/main...cpu:rustls:cpu-quic-support-review) is this branch with review feedback applied and the last commit switched out for the `Tls13CipherSuite` update I described above.
* [`cpu-rustls-0.23`](https://github.com/quinn-rs/quinn/compare/main...cpu:quinn:cpu-rustls-0.23) is https://github.com/quinn-rs/quinn/pull/1715 updated to:
  * Track this branch, and rustls 0.23.0-alpha.0.
  * Use the new updated `Tls13CipherSuite` and `quic::Suite` types.
  
Like quinn#1715 the `quinn-proto` tests are passing, but there's more work to do in the other crates.

Reply:
> there's more work to do in the other crates.

[`cpu-rustls-0.23-all-crates`](https://github.com/quinn-rs/quinn/compare/main...cpu:quinn:cpu-rustls-0.23-all-crates) updates the other crates so they'll build (note the `quinn` tests aren't passing yet).

For the `bench` and `perf` crates I switched what was the implied crypto provider of `ring` to `aws_lc_rs`.

I left the main `quinn` crate using the process-wide default provider, but it means the unit tests fail because there's no provider set. I think there are a few options to figure this out but I'm going to pause here for input on the Rustls branch changes and the direction you want to see Quinn go w.r.t default crypto.

Reply:
> is this branch with review feedback applied and the last commit switched out for the Tls13CipherSuite update I described above.

I like this, FWIW

Reply:
> I like this, FWIW

SG. I'll push the update and we can look at the Quinn side separately.

Reply:
> I'll push the update

(Done)

Reply:
Here's [a branch](https://github.com/cpu/quinn/tree/cpu-rustls-0.23-all-crates-ring-default) that passes green in CI using this branch of Rustls (minus a FreeBSD task complaining about build deps - haven't looked into this).

It's sufficient for showing Quinn can take this branch without error, but would need further cleaning up to upstream.